pkg/apis/resourcedefinition/helper.go

```diff
@@ -0,0 +1,528 @@
+
+	for _, k := range []string{variablesSchemaKey, outputsSchemaKey} {
+		var lsr *openapi3.SchemaRef
+
+		for _, s := range scs {
+			rsr := s[k]
+
+			// If the schema ref is not object,
+			// reset and skip.
+			if rsr == nil || rsr.Value == nil || len(rsr.Value.Properties) == 0 {
+				lsr = nil
+				break
+			}
+
+			// Remove original extension.
+			for _, p := range rsr.Value.Properties {
+				if p.Value == nil || len(p.Value.Extensions) == 0 {
+					continue
+				}
+
+				delete(p.Value.Extensions, originalExtensionKey)
+			}
+
+			// If the schema ref is nil,
+			// reuse and go ahead.
+			if lsr == nil {
+				lsr = rsr
+				continue
+			}
+
+			// Arrange the schema ref.
+			alignSchemaRef(lsr, rsr)
+		}
+
+		if lsr != nil && len(lsr.Value.Properties) != 0 {
+			ret[k] = lsr
+		}
+	}
+
+	return ret
+}
+
+// alignSchemaRef aligns the schema,
+// which always move in the direction of shrinking at first, if not found, then expanding.
+func alignSchemaRef(lsr, rsr *openapi3.SchemaRef) {
+	if lsr == nil || lsr.Value == nil ||
+		rsr == nil || rsr.Value == nil {
+		return
+	}
+
+	lv, rv := lsr.Value, rsr.Value
+
+	switch lv.Type {
+	case openapi3.TypeObject:
+		if lv.Properties != nil {
+			reqs := sets.NewString(lv.Required...)
+
+			for k := range lv.Properties {
+				// Delete the key if the type is not the same.
+				if !isSchemaRefTypeEqual(lv.Properties[k], rv.Properties[k]) {
+					delete(lv.Properties, k)
+					reqs.Delete(k)
+
+					continue
+				}
+
+				alignSchemaRef(lv.Properties[k], rv.Properties[k])
+			}
+
+			// Reset required keys.
+			if reqs.Len() == 0 {
+				lsr.Value.Required = nil
+			} else {
+				lsr.Value.Required = reqs.List()
+			}
+		} else {
+			alignSchemaRef(lv.AdditionalProperties.Schema, rv.AdditionalProperties.Schema)
+		}
+	case openapi3.TypeArray:
+		alignSchemaRef(lv.Items, rv.Items)
+	}
+
+	if lv.Format == "" && rv.Format != "" {
+		lv.Format = rv.Format
+	}
+
+	if !lv.WriteOnly && rv.WriteOnly {
+		lv.WriteOnly = rv.WriteOnly
+	}
+
+	// Find the most complete sentence.
+	// If not found, it falls back to the recommended description.
+	if mef := "description"; !isSchemaRefMutuallyExclusive(lsr, mef) {
+		switch {
+		case lv.Description == "" && rv.Description != "":
+			lv.Description = rv.Description
+		case lv.Description != "" && rv.Description != "":
+			var (
+				d      []rune
+				period int
+			)
+			lvd, rvd := []rune(lv.Description), []rune(rv.Description)
+			s := int(math.Min(float64(len(lvd)), float64(len(rvd))))
+
+			for i := 0; i < s; i++ {
+				if lvd[i] != rvd[i] {
+					break
+				}
+
+				d = append(d, lvd[i])
+
+				if lvd[i] == '.' {
+					period = i
+				}
+			}
+
+			if period != 0 {
+				lv.Description = string(d)
+				if period+1 < len(d) {
+					lv.Description = string(d[:period+1])
+				}
+			} else {
+				var rt string
+				if t := strings.ToLower(lv.Title); t != strs.Pluralize(t) {
+					rt = "Specify the " + t + "."
+				} else {
+					rt = "Specify " + t + "."
+				}
```

Comment:
The logic depending the peroid seems fragile. What if the description uses other terminal punctuation like `?/!` or anthing else? Is there a more universal approach, like getting the longest common substring?